File: my-folder/0236-lowest-common-ancestor-of-a-binary-tree/solution.java
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) {
          return null;
        }
        
        return lca(root, p, q);
    }
    
    private TreeNode lca(TreeNode node, TreeNode p, TreeNode q) {
      if (node.left == null && node.right == null) {
        return node == p || node == q ? node : null;
      }
      
      boolean isNodePresent = node == p || node == q;
      
      TreeNode leftLca = null;
      TreeNode rightLca = null;
      
      if (node.left != null) {
        leftLca = lca(node.left, p, q);
      }
      
      if (node.right != null) {
        rightLca = lca(node.right, p, q);
      }
      
      if (leftLca != null && rightLca != null) {
        return node;
      }
      
      if (isNodePresent) {
        return node;
      }
      
      return leftLca != null ? leftLca : rightLca;
    }
}

File: my-folder/0236-lowest-common-ancestor-of-a-binary-tree/solution.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int = 0) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
  lateinit var lca: TreeNode
    
    fun lowestCommonAncestor(root: TreeNode?, p: TreeNode?, q: TreeNode?): TreeNode? {
      isNodePresent(root, p, q)
      
      return lca
    }
    
    fun isNodePresent(
      node: TreeNode?,
      p: TreeNode?,
      q: TreeNode?
    ): Boolean {
      if (node == null) return false
      
      val atNode = if (node.`val` == p?.`val` || node.`val` == q?.`val`) true else false
      val left = isNodePresent(node.left, p, q)
      val right = isNodePresent(node.right, p, q)
      
      if (left && right) { lca = node }
      if (atNode && left || atNode && right) { lca = node }
      
      return if (atNode || left || right) true else false
    }
}
